engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnJobActivationBehavior.java

```diff
@@ -28,26 +38,70 @@
-  private final VariableState variableState;
+  private final JobVariablesCollector jobVariablesCollector;
   private final StateWriter stateWriter;
   private final SideEffectWriter sideEffectWriter;
+  private final KeyGenerator keyGenerator;
   private final JobMetrics jobMetrics;
 
   public BpmnJobActivationBehavior(
       final JobStreamer jobStreamer,
       final VariableState variableState,
       final Writers writers,
+      final KeyGenerator keyGenerator,
       final JobMetrics jobMetrics) {
     this.jobStreamer = jobStreamer;
-    this.variableState = variableState;
+    this.keyGenerator = keyGenerator;
+    this.jobMetrics = jobMetrics;
+    jobVariablesCollector = new JobVariablesCollector(variableState);
     stateWriter = writers.state();
     sideEffectWriter = writers.sideEffect();
-    this.jobMetrics = jobMetrics;
   }
 
-  public void publishWork(final JobRecord jobRecord) {
-    final String jobType = jobRecord.getType();
-    notifyJobAvailable(jobType);
+  public void publishWork(final long jobKey, final JobRecord jobRecord) {
+    final JobRecord wrappedJobRecord = new JobRecord();
+    wrappedJobRecord.wrapWithoutVariables(jobRecord);
+
+    final String jobType = wrappedJobRecord.getType();
+    final Optional<JobStream> optionalJobStream =
+        jobStreamer.streamFor(wrappedJobRecord.getTypeBuffer());
+
+    if (optionalJobStream.isPresent()) {
+      final JobStream jobStream = optionalJobStream.get();
+      final JobActivationProperties properties = jobStream.properties();
+
+      // we push the job immediately, so the deadline is always calculated from the current time
+      final var deadline = ActorClock.currentTimeMillis() + properties.timeout();
+      wrappedJobRecord.setDeadline(deadline);
+      wrappedJobRecord.setWorker(properties.worker());
+
+      // reuse the existing JobBatch activation mechanism
+      final JobBatchRecord jobBatchRecord = new JobBatchRecord();
+      jobBatchRecord
+          .setType(jobType)
+          .setTimeout(properties.timeout())
+          .setWorker(properties.worker());
+      appendJobToBatch(jobBatchRecord, jobKey, wrappedJobRecord);
+      final var jobBatchKey = keyGenerator.nextKey();
+      stateWriter.appendFollowUpEvent(jobBatchKey, JobBatchIntent.ACTIVATED, jobBatchRecord);
+
+      jobVariablesCollector.setJobVariables(
+          properties.fetchVariables(), wrappedJobRecord, wrappedJobRecord.getElementInstanceKey());
+      final var pushableJobRecord = new JobRecord();
+      cloneJob(wrappedJobRecord, pushableJobRecord);
+      final var activatedJob = new ActivatedJobImpl();
+      activatedJob.setJobKey(jobKey).setRecord(pushableJobRecord);
+
+      // job push through side effect
+      sideEffectWriter.appendSideEffect(
+          () -> {
+            jobStream.push(activatedJob);
+            jobMetrics.jobPush(jobType);
+            return true;
+          });
+    } else {
+      notifyJobAvailable(jobType);
+    }
   }
```

Comment:
🔧 Above is only a refactoring change to make work publishing more readable. No logic change.

Reply:
@berkaycanbc I'm not sure that the `wrapJobRecordWithoutVariables`, `createPushableJob`, and `createActivatedJob` methods really improve readability. It might even be less readable since you need to go back-and-forth between methods to understand the flow of the code.

Do you think we really need to refactor that code?

Reply:
@koevskinikola Yes, I think it is needed. But at the same time, as long as we split the `publishWork` method into reasonable steps, it should just work. The latests changes you just send is totally sufficient! 🚀